File: fn/src/main/java/org/sqlbuilder/fn/joins/Pair.java
package org.sqlbuilder.fn.joins;

import java.util.Comparator;
import java.util.Objects;

public class Pair<T, U>
{
	protected final T first;

	protected final U second;

	public Pair(final T first, final U second)
	{
		this.first = first;
		this.second = second;
	}

	public T getFirst()
	{
		return first;
	}

	public U getSecond()
	{
		return second;
	}

	@Override
	public boolean equals(final Object o)
	{
		if (this == o)
		{
			return true;
		}
		if (o == null || getClass() != o.getClass())
		{
			return false;
		}
		Pair<?, ?> that = (Pair<?, ?>) o;
		return Objects.equals(first, that.first) && Objects.equals(second, that.second);
	}

	@Override
	public int hashCode()
	{
		return Objects.hash(first, second);
	}

	@Override
	public String toString()
	{
		return String.format("[VU-AS first=%s second=%s]", first, second);
	}
}

File: fn/src/main/java/org/sqlbuilder/fn/joins/Pair.kt
package org.sqlbuilder.fn.joins

import java.util.*

open class Pair<T, U>(
    val first: T,
    val second: U) {

    override fun equals(o: Any?): Boolean {
        if (this === o) {
            return true
        }
        if (o == null || javaClass != o.javaClass) {
            return false
        }
        val that = o as Pair<*, *>
        return first == that.first && second == that.second
    }

    override fun hashCode(): Int {
        return Objects.hash(first, second)
    }

    override fun toString(): String {
        return String.format("[first=$first second=$second]")
    }
}
